packages/flutter_tools/lib/src/ios/fallback_discovery.dart

```diff
@@ -0,0 +1,126 @@
+///
+/// Finally, if neither of the above approaches works, we can still attempt
+/// to parse logs.
+///
+/// To improve the overall resilience of the process, this class combines the
+/// three discovery strategies. First it assigns a port and attempts to connect.
+/// Then if this fails it falls back to mDNS, then finally attempting to scan
+/// logs.
+class FallbackDiscovery {
+  FallbackDiscovery({
+    @required DevicePortForwarder portForwarder,
+    @required MDnsObservatoryDiscovery mDnsObservatoryDiscovery,
+    @required Logger logger,
+    @required ProtocolDiscovery protocolDiscovery,
+    @required HttpClient httpClient,
+  }) : _logger = logger,
+       _mDnsObservatoryDiscovery = mDnsObservatoryDiscovery,
+       _portForwarder = portForwarder,
+       _protocolDiscovery = protocolDiscovery,
+       _httpClient = httpClient;
+
+  final DevicePortForwarder _portForwarder;
+  final MDnsObservatoryDiscovery _mDnsObservatoryDiscovery;
+  final Logger _logger;
+  final ProtocolDiscovery _protocolDiscovery;
+  final HttpClient _httpClient;
+
+  /// Attempt to discover the observatory port.
+  Future<Uri> discover({
+    @required int assumedDevicePort,
+    @required String packageId,
+    @required Device deivce,
+    @required bool usesIpv6,
+    @required int hostVmservicePort,
+  }) async {
+    try {
+      final int hostPort = await _portForwarder.forward(assumedDevicePort, hostPort: hostVmservicePort);
+      final Uri assumedUri = Uri.parse('http://localhost:$hostPort');
+      // Attempt to connect to the vmservice.
+      int attempts = 0;
+      const int delaySeconds = 2;
+      while (attempts < 5) {
+        // Making a GET request to the forwarded URL should return an HTML
+        // response page containing a specific string. This logic needs to
+        // be tightened up to ensure we don't accidentally try to connect
+        // to some arbitrary server.
+        final HttpClientRequest request = await _httpClient.getUrl(assumedUri);
+        final HttpClientResponse response = await request.close();
+        if (response.statusCode == HttpStatus.ok) {
+          final String responseBody = await response.transform(utf8.decoder).join('');
+          if (responseBody.contains('Dart')) {
+            _httpClient.close();
+            UsageEvent('ios-mdns', 'precheck-success').send();
+            return assumedUri;
+          }
+        }
+
+        // No exponential backoff is used here to keep the amount of time the
+        // tool waits for a connection to be reasonable. If the vmservice cannot
+        // be connected to in this way, the mDNS discovery must be reached
+        // sooner rather than later.
+        await Future<void>.delayed(const Duration(seconds: delaySeconds));
+        attempts += 1;
+      }
+    } on Exception {
+      _logger.printTrace('Failed to connect directly, falling back to mDNS');
+    }
+    UsageEvent('ios-mdns', 'precheck-failure').send();
```

Comment:
Consider creating a new event type in events.dart. It might also be interesting to record the port number that we tried for here in case there's one that fails a lot.

Reply:
I created ios-handshake as an event type. I don't think it would be worth it to record the port, since we're effectively choosing a random number, we'd need a tremendous amount of data to get any insight into the failures.